user.py

```diff
@@ -32,33 +48,84 @@ def add_user():
     name = request.form['name']
     first_name = request.form['first_name']
     email = request.form['email']
-
     if not contains_valid_characters(name):
         return make_response("Invalid characters in name", 400)
     if not contains_valid_characters(first_name):
         return make_response("Invalid characters in first name", 400)
     if email != '' and not is_valid_email(email):
         return make_response("Invalid email format", 400)
 
+    organization_code = request.form['organization_code']
+    is_teacher = 'is_teacher' in request.form
+    is_researcher = 'is_researcher' in request.form
+    supervisor_id = request.form.get('supervisor') if is_researcher else None
+    researcher_type = request.form['researcher_type'] if is_researcher else None
+
     try:
-        new_user = User(name=name, first_name=first_name, email=email)
+        new_user = User(name=name, first_name=first_name, email=email, is_teacher=is_teacher,
+                        is_researcher=is_researcher, supervisor_id=supervisor_id,
+                        organization_id=organization_code)
         db.session.add(new_user)
         db.session.commit()
+        if is_researcher:
+            max_load = get_default_max_load(researcher_type)
+            new_researcher = Researcher(user_id=new_user.id, researcher_type=researcher_type,
+                                        max_loads=max_load)
+            db.session.add(new_researcher)
+            db.session.commit()
     except:
         db.session.rollback()
         raise
-    return redirect(url_for("index"))
+
+    return redirect(url_for("user.register"))
 
 
-@user_bp.route('/profile')
+@user_bp.route('/users/<string:user_type>')
 @login_required
-def profile():
-    return render_template("profile.html")
+def users(user_type):
+    base_query = db.session.query(User).filter(User.admin == 0)
+
+    if user_type == 'teacher':
+        base_query = base_query.filter(User.is_teacher == 1, User.active == 1)
+    elif user_type == 'researcher':
+        base_query = base_query.filter(User.is_researcher == 1, User.active == 1)
+    elif user_type == 'archived':
+        base_query = base_query.filter(User.active == 0)
+
+    all_users = base_query.all()
+    return render_template('users.html', users=all_users, user_type=user_type)
 
 
-@user_bp.route('update_profile', methods=['POST'])
+@user_bp.route('/profile/<int:user_id>')
 @login_required
-def update_profile():
+def user_profile(user_id):
+    all_users = db.session.query(User).filter(User.admin == 0, User.is_teacher == 1, User.active == 1).all()
+    my_user = db.session.query(User).filter_by(id=user_id).first()
```

Comment:
Maybe `my_user` is a bad name as it contains the requested user, while `my_profile` is really session-specific.